### PR TITLE
remove THHalf and use at::Half instead

### PR DESCRIPTION
**Pull request by li-roy**

### CODE REVIEW

**Comment by ezyang** [comment]
I'd approve you, but you're gonna have to get the builds to pass first ;)

**Comment by gchanan** [comment]
I may be running into problems because we try to compile TH.h as C code to check that we don't break the C FFI.  So far we've avoided breaking that when we can.  This might be tricky with THHalf because there may be a point where you really need the definition in the .h, and you can't do that because of the namespace.  It might be possible to reorganize the code to avoid that, though.

**Comment by colesbury** [comment]
We should stop checking for C compatibility. I don't think it's worth it at this point.

**Comment by gchanan** [comment]
I don't disagree, just pointing it out (and also that I don't know how difficult it is to remove the assumptions necessary to get the build passing).

**Comment by colesbury** [comment]
Greg pointed out that you can also use the same trick as `generic/THTensor.h`:

```c
#ifdef __cplusplus
#define THTensor at::TensorImpl
#else
typedef struct at_Tensor_Impl at_Tensor_Impl;
#define THTensor at_Tensor_Impl
#endif
```

(Substitute THHalf for THTensor and at::Half for at::TensorImpl)

That would keep the THHalf name, but not as a distinct type.

File: .jenkins/pytorch/build.sh [hunk 1]
```diff
@@ -93,17 +93,6 @@ if [[ "$BUILD_ENVIRONMENT" != *cuda* ]]; then
   time tools/test_aten_install.sh
 fi
 
-# Test C FFI plugins
```
Comment (ezyang) [inline .jenkins/pytorch/build.sh]: This removal needs to be noted prominently in the commit message

**Comment by ezyang** [comment]
If ffi is not working in OSS, I'm skeptical that all fbcode builds will still be working...

**Comment by li-roy** [comment]
closed in favor of #11010, will deal cffi issues in a future pr.

**Pull request closed without merging**